### PR TITLE
Fix: Re-importing edge case

### LINKED ISSUE #28: Potential race condition when loading a URL multiple times
https://github.com/Polymer/lazy-imports/blob/master/lazy-imports-import.html#L34

We set the bit on whether a URL is loaded synchronously, before the result of loading it is actually known. A subsequent call to importLazyGroup with the same URL could report that the URL is loaded before it has.

### PR DESCRIPTION
**Pull request by stramel**
Following up from @justinfagnani's comments on #29. Fixes #28 and #30 

May be a bit easier to review without the whitespace, https://github.com/Polymer/lazy-imports/pull/36/files?w=1

**TODO:**

- [x] Tests
- [x] Documentation

### CODE REVIEW

File: lazy-imports-import.html [hunk 1]
```diff
@@ -9,38 +9,58 @@
 -->
 <script>
   (function() {
-    var __loadedImports = {};
+    var __importPromises = {};
 
-    Polymer.__importLazyGroup = function(callingElement, group, fromElement) {
+    /**
+     * TODO:
```
Comment (justinfagnani) [inline lazy-imports-import.html]: todo :)

**Comment by justinfagnani** [comment]
Looks good except the one todo

**Comment by stramel** [comment]
@justinfagnani I missed a couple things with passing options through. I'm going to push that and address my TODOs that are outstanding.

**Comment by stramel** [comment]
@justinfagnani I have addressed everything and should be ready for review again. Thanks!

**Comment by stramel** [comment]
@justinfagnani I played around with it trying to clean it up a bit... came up with an alternative way. Let me know which you think is better! Also, let me know if you notice anyway to improve it.
https://gist.github.com/stramel/15c69eb58559d95523df00482c2dc60e

**Comment by stramel** [comment]
@justinfagnani Mind taking another look?

@rictic Do you want to take a look as well?

File: lazy-imports-import.html [hunk 2]
```diff
@@ -9,37 +9,71 @@
...
-            __loadedImports[resolved] = true;
+
+      var links =
+        Polymer.DomModule.import(fromElement.localName).querySelectorAll(query);
+
+      if (links.length < 1) {
+        return Promise.resolve(importStatuses);
+      }
+
+      var promises = Array.from(links).map(function(link) {
```
Comment (justinfagnani) [inline lazy-imports-import.html]: I like more specific names... what kind of promises?

File: lazy-imports-import.html [hunk 3]
```diff
@@ -9,37 +9,71 @@
...
 <script>
   (function() {
-    var __loadedImports = {};
+    'use strict';
 
-    Polymer.__importLazyGroup = function(callingElement, group, fromElement) {
-      if (!fromElement) {
-        fromElement = Polymer.DomModule.import(callingElement.localName);
-      }
+    var __importPromises = {};
```
Comment (justinfagnani) [inline lazy-imports-import.html]: Map is supported everywhere Polymer is, let's use that.

**Review by justinfagnani — APPROVED** [review]

File: lazy-imports-import.html (hunk 3 above)
Comment (stramel) [reply lazy-imports-import.html]: @justinfagnani Does that include Polymer 1.x? This is the core implementation for both the behavior and mixin. Previously have excluded any ES6 functionality and enforced through linter

Comment (justinfagnani) [reply lazy-imports-import.html]: IE11 has basic Map support, so it's ok

Comment (stramel) [reply lazy-imports-import.html]: Awesome! Thanks for the tip! Will get these changes made!

**Comment by stramel** [comment]
@justinfagnani Feedback addressed :+1:

**Comment by stramel** [comment]
:scream: :tada: Thanks!

**Pull request merged**